codec/enforceproto/extraneous_fields.go

```diff
@@ -0,0 +1,355 @@
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"reflect"
+	"sync"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/gogo/protobuf/protoc-gen-gogo/descriptor"
+	"google.golang.org/protobuf/encoding/protowire"
+
+	"github.com/cosmos/cosmos-sdk/codec/types"
+)
+
+const bit11NonCritical = 1 << 10
+
+type descriptorIface interface {
+	Descriptor() ([]byte, []int)
+}
+
+// CheckMismatchedProtoFields walks through the protobuf serialized bytes in b, and tries to
+// compare field numbers and wireTypes against what msg expects. The error returned if non-nil will contain
+// the listing of the extraneous fields by tagNumber and wireType, or mismatched wireTypes.
+func CheckMismatchedProtoFields(b []byte, msg proto.Message) error {
+	if len(b) == 0 {
+		return nil
+	}
+
+	desc, ok := msg.(descriptorIface)
+	if !ok {
+		return fmt.Errorf("%T does not have a Descriptor() method", msg)
+	}
+
+	fieldDescProtoFromTagNum, _, err := fieldNumToFieldDesc(desc, msg)
+	if err != nil {
+		return err
+	}
+
+	for len(b) > 0 {
+		tagNum, wireType, n := protowire.ConsumeField(b)
+		if n < 0 {
+			return errors.New("invalid length")
+		}
+
+		fieldDescProto, ok := fieldDescProtoFromTagNum[int32(tagNum)]
+		switch {
+		case ok:
+			// Assert that the wireTypes match.
+			if !canEncodeType(wireType, fieldDescProto.GetType()) {
+				return &errMismatchedWireType{
+					Type:         reflect.ValueOf(msg).Type().String(),
+					TagNum:       tagNum,
+					GotWireType:  wireType,
+					WantWireType: protowire.Type(fieldDescProto.WireType()),
+				}
+			}
+
+		default:
+			if tagNum&bit11NonCritical == 0 {
+				// The tag is non-critical, so report it.
+				return &errExtraneousField{
+					Type:     reflect.ValueOf(msg).Type().String(),
+					TagNum:   tagNum,
+					WireType: wireType,
+				}
+			}
+		}
+
+		// Skip over the 2 bytes that store fieldNumber and wireType bytes.
+		fieldBytes := b[2:n]
+		b = b[n:]
+
+		// An unknown but non-critical field or just a scalar type (aka *INT and BYTES like).
+		if fieldDescProto == nil || fieldDescProto.IsScalar() {
+			continue
+		}
+
+		protoMessageName := fieldDescProto.GetTypeName()
+		if protoMessageName == "" {
+			// At this point only TYPE_STRING is expected to be unregistered, since FieldDescriptorProto.IsScalar() returns false for TYPE_STRING
+			// per https://github.com/gogo/protobuf/blob/5628607bb4c51c3157aacc3a50f0ab707582b805/protoc-gen-gogo/descriptor/descriptor.go#L95-L118
+			if typ := fieldDescProto.GetType(); typ != descriptor.FieldDescriptorProto_TYPE_STRING {
+				return fmt.Errorf("failed to get typename for message of type %d, can only be TYPE_STRING", typ)
+			}
+			continue
+		}
+
+		// Let's recursively traverse and typecheck the field.
+
+		if protoMessageName == ".google.protobuf.Any" {
+			// We'll need to extract the TypeURL which will contain the protoMessageName.
+			any := new(types.Any)
+			if err := proto.Unmarshal(fieldBytes, any); err != nil {
+				return err
+			}
+			protoMessageName = any.TypeUrl
```

Comment:
I think you need to set fieldBytes here too. There must be a test missing or I'm misreading the logic

Reply:
You were correct, it would skip to it but was adding extra work, so this removes that problem.